File: ios/Classes/FlutterDaroRewardAd.swift
import Daro

public class FlutterDaroRewardAdLoadListener {
    var onAdLoadSuccess: ((_ adItem: FlutterDaroRewardAd, _ ad: Any?, _ adInfo: Any?) -> Void)?
    var onAdLoadFail: ((_ error: Error) -> Void)?
    var onAdImpression: ((_ adInfo: Any?) -> Void)?
    var onAdClicked: ((_ adInfo: Any?) -> Void)?
    
    public init(
        onAdLoadSuccess: ((_ adItem: FlutterDaroRewardAd, _ ad: Any?, _ adInfo: Any?) -> Void)? = nil,
        onAdLoadFail: ((_ error: Error) -> Void)? = nil,
        onAdImpression: ((_ adInfo: Any?) -> Void)? = nil,
        onAdClicked: ((_ adInfo: Any?) -> Void)? = nil
    ) {
        self.onAdLoadSuccess = onAdLoadSuccess
        self.onAdLoadFail = onAdLoadFail
        self.onAdImpression = onAdImpression
        self.onAdClicked = onAdClicked
    }
    
    deinit {
        onAdLoadSuccess = nil
        onAdLoadFail = nil
        onAdImpression = nil
        onAdClicked = nil
    }
}
public class FlutterDaroRewardAdListener {
    var onShown: ((_ adInfo: Any?) -> Void)?
    var onRewarded: ((_ adInfo: Any?, _ reward: DaroRewardedItem?) -> Void)?
    var onDismiss: ((_ adInfo: Any?) -> Void)?
    var onFailedToShow: ((_ adInfo: Any?, _ error: Error) -> Void)?
    
    public init(
        onShown: ((_ adInfo: Any?) -> Void)? = nil,
        onRewarded: ((_ adInfo: Any?, _ reward: DaroRewardedItem?) -> Void)? = nil,
        onDismiss: ((_ adInfo: Any?) -> Void)? = nil,
        onFailedToShow: ((_ adInfo: Any?, _ error: Error) -> Void)? = nil
    ) {
        self.onShown = onShown
        self.onRewarded = onRewarded
        self.onDismiss = onDismiss
        self.onFailedToShow = onFailedToShow
    }
    
    deinit {
        onShown = nil
        onRewarded = nil
        onDismiss = nil
        onFailedToShow = nil
    }
}

// 광고 타입 정의
public enum FlutterDaroRewardAdType: String {
    case interstitial
    case rewardedVideo
    case popup
    case opening
}

public class FlutterDaroRewardAd: UIViewController {
    private var adType: FlutterDaroRewardAdType
    private var adUnit: String
    private var placement: String?
    private var loader: Any?
    private var ad: Any?

    // 광고 리스너
    var loadListener: FlutterDaroRewardAdLoadListener?
    
    public init(
        adType: FlutterDaroRewardAdType, 
        adUnit: String, 
        placement: String? = nil, 
        listener: FlutterDaroRewardAdLoadListener? = nil) {
            self.adType = adType
            self.adUnit = adUnit
            self.placement = placement
            self.loadListener = listener
            super.init(nibName: nil, bundle: nil)
            setupLoader()
    }

    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    public override func viewDidLoad() {
        super.viewDidLoad()
    }

    // 광고 타입에 따라 알맞은 로더 인스턴스 생성
    private func setupLoader() {
        switch adType {
        case .interstitial:
            let unit = DaroAdUnit(unitId: adUnit)
            self.loader = DaroInterstitialAdLoader(unit: unit)
        case .rewardedVideo:
            let unit = DaroAdUnit(unitId: adUnit)
            self.loader = DaroRewardedAdLoader(unit: unit)
        case .popup:
            let unit = DaroAdUnit(unitId: adUnit)
            self.loader = DaroLightPopupAdLoader(unit: unit)
        case .opening:
            let unit = DaroAdUnit(unitId: adUnit)
            self.loader = DaroAppOpenAdLoader(unit: unit)
        }
    }
    
    // 광고 로드 실행
    public func loadAd(autoShow: Bool = false, listener: FlutterDaroRewardAdListener? = nil, result: @escaping (Bool, Error?) -> Void) {
        if self.loader == nil {
            self.setupLoader()
        }
        if let interstitialLoader = self.loader as? DaroInterstitialAdLoader {
            interstitialLoader.listener.onAdLoadSuccess = { [weak self] ad, adInfo in
                self?.ad = ad
                self?.loadListener?.onAdLoadSuccess?(self!, ad, adInfo)
                // 즉시 노출 옵션이 true일 경우 광고 표시
                if autoShow {
                    self?.showAd(listener: listener) { success, error in 
                        result(success, error)
                    }
                } else {
                    result(true, nil)
                }
            }
            interstitialLoader.listener.onAdLoadFail = { [weak self] error in
                self?.loadListener?.onAdLoadFail?(error)
                result(false, error)
            }
            interstitialLoader.listener.onAdImpression = { [weak self] adInfo in
                self?.loadListener?.onAdImpression?(adInfo)
            }
            interstitialLoader.listener.onAdClicked = { [weak self] adInfo in
                self?.loadListener?.onAdClicked?(adInfo)
            }
            interstitialLoader.loadAd()
        } else if let rewardedLoader = self.loader as? DaroRewardedAdLoader {
            rewardedLoader.listener.onAdLoadSuccess = { [weak self] ad, adInfo in
                self?.ad = ad
                self?.loadListener?.onAdLoadSuccess?(self!, ad, adInfo)
                // 즉시 노출 옵션이 true일 경우 광고 표시
                if autoShow {
                    self?.showAd(listener: listener) { success, error in 
                        result(success, error)
                    }
                } else {
                    result(true, nil)
                }
            }
            rewardedLoader.listener.onAdLoadFail = { [weak self] error in
                self?.loadListener?.onAdLoadFail?(error)
                result(false, error)
            }
            rewardedLoader.listener.onAdImpression = { [weak self] adInfo in
                self?.loadListener?.onAdImpression?(adInfo)
            }
            rewardedLoader.listener.onAdClicked = { [weak self] adInfo in
                self?.loadListener?.onAdClicked?(adInfo)
            }

            rewardedLoader.loadAd()
        } else if let popupLoader = self.loader as? DaroLightPopupAdLoader {
            popupLoader.listener.onAdLoadSuccess = { [weak self] ad, adInfo in
                self?.ad = ad
                self?.loadListener?.onAdLoadSuccess?(self!, ad, adInfo)
                // 즉시 노출 옵션이 true일 경우 광고 표시
                if autoShow {
                    self?.showAd(listener: listener) { success, error in 
                        result(success, error)
                    }
                } else {
                    result(true, nil)
                }
            }
            popupLoader.listener.onAdLoadFail = { [weak self] error in
                self?.loadListener?.onAdLoadFail?(error)
                result(false, error)
            }
            popupLoader.listener.onAdImpression = { [weak self] adInfo in
                self?.loadListener?.onAdImpression?(adInfo)
            }
            popupLoader.listener.onAdClicked = { [weak self] adInfo in
                self?.loadListener?.onAdClicked?(adInfo)
            }
            
            popupLoader.loadAd()
        } else if let openingLoader = self.loader as? DaroAppOpenAdLoader {
            openingLoader.listener.onAdLoadSuccess = { [weak self] ad, adInfo in
                self?.ad = ad
                self?.loadListener?.onAdLoadSuccess?(self!, ad, adInfo)
                // 즉시 노출 옵션이 true일 경우 광고 표시
                if autoShow {
                    self?.showAd(listener: listener) { success, error in 
                        result(success, error)
                    }
                } else {
                    result(true, nil)
                }
            }
            openingLoader.listener.onAdLoadFail = { [weak self] error in
                self?.loadListener?.onAdLoadFail?(error)
                result(false, error)
            }
            openingLoader.listener.onAdImpression = { [weak self] adInfo in
                self?.loadListener?.onAdImpression?(adInfo)
            }
            openingLoader.listener.onAdClicked = { [weak self] adInfo in
                self?.loadListener?.onAdClicked?(adInfo)
            }
            openingLoader.loadAd()
        } else {
            result(false, NSError(domain: "FlutterDaroRewardAd", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid ad type"]))
        }
    }

    // 광고 표시 처리
    public func showAd(listener: FlutterDaroRewardAdListener?, result: @escaping (Bool, Error?) -> Void) {
        // rootViewController 가져오기
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
            result(false, NSError(domain: "FlutterDaroRewardAd", code: -1, userInfo: [NSLocalizedDescriptionKey: "No root view controller available"]))
            return
        }
        
        // 광고 타입별로 실제 광고 표시 함수 호출 및 리스너 연결
        if let interstitialAd = self.ad as? DaroInterstitialAd {
            interstitialAd.interstitialListener.onShown = { adInfo in
                listener?.onShown?(adInfo)
            }
            interstitialAd.interstitialListener.onDismiss = { adInfo in
                listener?.onDismiss?(adInfo)
                result(true, nil)
            }
            interstitialAd.interstitialListener.onFailedToShow = { adInfo, error in
                listener?.onFailedToShow?(adInfo, error)
                result(false, error)
            }
            // Interstitial 광고 표시 - rootViewController 사용
            interstitialAd.show(viewController: rootViewController)
        } else if let rewardedAd = self.ad as? DaroRewardedAd {
            var reward: DaroRewardedItem?
            rewardedAd.rewardedAdListener.onShown = { adInfo in
                listener?.onShown?(adInfo)
            }
            rewardedAd.rewardedAdListener.onEarnedReward = { adInfo, rewardedItem in
                listener?.onRewarded?(adInfo, rewardedItem)
                reward = rewardedItem
            }
            rewardedAd.rewardedAdListener.onDismiss = { adInfo in
                listener?.onDismiss?(adInfo)
                result(reward != nil, nil)
            }
            rewardedAd.rewardedAdListener.onFailedToShow = { adInfo, error in
                listener?.onFailedToShow?(adInfo, error)
                result(false, error)
            }
            // Rewarded Video 광고 표시 - rootViewController 사용
            rewardedAd.show(viewController: rootViewController)
        } else if let popupAd = self.ad as? DaroLightPopupAd {
            popupAd.lightPopupAdListener.onShown = { adInfo in
                listener?.onShown?(adInfo)
            }
            popupAd.lightPopupAdListener.onDismiss = { adInfo in
                listener?.onDismiss?(adInfo)
                result(true, nil)
            }
            popupAd.lightPopupAdListener.onFailedToShow = { adInfo, error in
                listener?.onFailedToShow?(adInfo, error)
                result(false, error)
            }
            // Popup 광고 표시 - rootViewController 사용
            popupAd.show(viewController: rootViewController)
        } else if let openingAd = self.ad as? DaroAppOpenAd {
            openingAd.appOpenAdListener.onShown = { adInfo in
                listener?.onShown?(adInfo)
            }
            openingAd.appOpenAdListener.onDismiss = { adInfo in
                listener?.onDismiss?(adInfo)
            }
            openingAd.appOpenAdListener.onFailedToShow = { adInfo, error in
                listener?.onFailedToShow?(adInfo, error)
            }
            // Opening 광고 표시
            openingAd.show()
        } else {
            self.loadAd(autoShow: true, listener: listener, result: result)
        }
    }

    deinit {
        self.destroy()
    }

    public override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        self.destroy()
    }

    public func destroy() {
        removeListeners()
        self.loader = nil
        self.ad = nil
    }

    private func removeListeners() {
        // 리스너 nil 처리로 강한 참조 해제
        if let interstitialAd = ad as? DaroInterstitialAd {
            interstitialAd.interstitialListener.onShown = nil
            interstitialAd.interstitialListener.onDismiss = nil
            interstitialAd.interstitialListener.onFailedToShow = nil
        } else if let rewardedAd = ad as? DaroRewardedAd {
            rewardedAd.rewardedAdListener.onShown = nil
            rewardedAd.rewardedAdListener.onEarnedReward = nil
            rewardedAd.rewardedAdListener.onDismiss = nil
            rewardedAd.rewardedAdListener.onFailedToShow = nil
        } else if let popupAd = ad as? DaroLightPopupAd {
            popupAd.lightPopupAdListener.onShown = nil
            popupAd.lightPopupAdListener.onDismiss = nil
            popupAd.lightPopupAdListener.onFailedToShow = nil
        } else if let openingAd = ad as? DaroAppOpenAd {
            openingAd.appOpenAdListener.onShown = nil
            openingAd.appOpenAdListener.onDismiss = nil
            openingAd.appOpenAdListener.onFailedToShow = nil
        }
        // loader 타입별로 리스너 해제
        if let interstitialLoader = loader as? DaroInterstitialAdLoader {
            interstitialLoader.listener.onAdLoadSuccess = nil
            interstitialLoader.listener.onAdLoadFail = nil
            interstitialLoader.listener.onAdClicked = nil
            interstitialLoader.listener.onAdImpression = nil
        } else if let rewardedLoader = loader as? DaroRewardedAdLoader {
            rewardedLoader.listener.onAdLoadSuccess = nil
            rewardedLoader.listener.onAdLoadFail = nil
            rewardedLoader.listener.onAdClicked = nil
            rewardedLoader.listener.onAdImpression = nil
        } else if let popupLoader = loader as? DaroLightPopupAdLoader {
            popupLoader.listener.onAdLoadSuccess = nil
            popupLoader.listener.onAdLoadFail = nil
            popupLoader.listener.onAdClicked = nil
            popupLoader.listener.onAdImpression = nil
        } else if let openingLoader = loader as? DaroAppOpenAdLoader {
            openingLoader.listener.onAdLoadSuccess = nil
            openingLoader.listener.onAdLoadFail = nil
            openingLoader.listener.onAdClicked = nil
            openingLoader.listener.onAdImpression = nil
        }
        self.loadListener = nil
    }
}

File: flutter-daro-sdk-a/android/src/main/kotlin/com/example/flutter_daro_sdk/FlutterDaroRewardAd.kt
package com.example.flutter_daro_sdk

import android.app.Activity
import android.content.Context
import android.graphics.Color
import androidx.core.graphics.toColorInt
import droom.daro.core.adunit.DaroAppOpenAdUnit
import droom.daro.core.adunit.DaroInterstitialAdUnit
import droom.daro.core.adunit.DaroLightPopupAdUnit
import droom.daro.core.adunit.DaroRewardedAdUnit
import droom.daro.core.listener.DaroAppOpenAdListener
import droom.daro.core.listener.DaroAppOpenAdLoaderListener
import droom.daro.core.listener.DaroInterstitialAdListener
import droom.daro.core.listener.DaroInterstitialAdLoaderListener
import droom.daro.core.listener.DaroLightPopupAdListener
import droom.daro.core.listener.DaroLightPopupAdLoaderListener
import droom.daro.core.listener.DaroRewardedAdListener
import droom.daro.core.listener.DaroRewardedAdLoaderListener
import droom.daro.core.model.DaroAdDisplayFailError
import droom.daro.core.model.DaroAdInfo
import droom.daro.core.model.DaroAdLoadError
import droom.daro.core.model.DaroAppOpenAd
import droom.daro.core.model.DaroInterstitialAd
import droom.daro.core.model.DaroLightPopupAd
import droom.daro.core.model.DaroLightPopupAdOptions
import droom.daro.core.model.DaroRewardedAd
import droom.daro.loader.DaroAppOpenAdLoader
import droom.daro.loader.DaroInterstitialAdLoader
import droom.daro.loader.DaroLightPopupAdLoader
import droom.daro.loader.DaroRewardedAdLoader

// 광고 타입 정의
enum class FlutterDaroRewardAdType(val value: String) {
    INTERSTITIAL("interstitial"),
    REWARDED_VIDEO("rewardedVideo"),
    POPUP("popup"),
    OPENING("opening");

    companion object {
        fun fromString(value: String?): FlutterDaroRewardAdType? {
            if(value == null) return null
            return entries.find { it.value == value }
        }
    }
}

// 광고 로드 리스너
interface FlutterDaroRewardAdLoadListener {
    fun onAdLoadSuccess(ad: FlutterDaroRewardAd, adInstance: Any?, adInfo: DaroAdInfo?) {}
    fun onAdLoadFail(error: DaroAdLoadError) {}
}

// 광고 표시 리스너
interface FlutterDaroRewardAdListener {
    fun onShown(adInfo: DaroAdInfo) {}
    fun onRewarded(adInfo: DaroAdInfo, rewardItem: DaroRewardedAd.DaroRewardedItem) {}
    fun onAdImpression(adInfo: DaroAdInfo) {}
    fun onAdClicked(adInfo: DaroAdInfo) {}
    fun onDismiss(adInfo: DaroAdInfo) {}
    fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {}
}

// Factory 클래스
object FlutterDaroRewardAdFactory {
    fun create(
        context: Context,
        adType: FlutterDaroRewardAdType,
        adUnit: String,
        placement: String? = null,
        options: Map<*,*>? = null,
        loadListener: FlutterDaroRewardAdLoadListener? = null
    ): FlutterDaroRewardAd {
        return when (adType) {
            FlutterDaroRewardAdType.INTERSTITIAL -> {
                FlutterDaroInterstitialAd(context, adUnit, placement, loadListener)
            }
            FlutterDaroRewardAdType.REWARDED_VIDEO -> {
                FlutterDaroRewardedVideoAd(context, adUnit, placement, loadListener)
            }
            FlutterDaroRewardAdType.POPUP -> {
                FlutterDaroPopupAd(context, adUnit, placement, loadListener, options)
            }
            FlutterDaroRewardAdType.OPENING -> {
                FlutterDaroOpeningAd(context, adUnit, placement, loadListener)
            }
        }
    }
}

// 추상 광고 클래스
abstract class FlutterDaroRewardAd(
    protected val context: Context,
    protected val adUnit: String,
    protected val placement: String? = null,
    protected var loadListener: FlutterDaroRewardAdLoadListener? = null
) {
    // 상위 클래스에서 공통으로 관리하는 loader와 ad
    private var loader: Any? = null
    protected var ad: Any? = null

    // 각 구체 클래스에서 로더 생성 구현
    protected abstract fun createLoader(): Any

    // 각 구체 클래스에서 로더 구현
    protected abstract fun loadAdInternal(
        loader: Any,
        autoShow: Boolean,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    )

    // 각 구체 클래스에서 광고 리스너 설정 및 표시 구현
    protected abstract fun showAdInternal(
        ad: Any,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    )

    // 각 구체 클래스에서 광고 해제 구현
    protected abstract fun destroyAd(ad: Any?)

    private fun setupLoader() {
        loader = createLoader()
    }

    // 공통 로드 로직
    fun loadAd(
        autoShow: Boolean = false,
        listener: FlutterDaroRewardAdListener? = null,
        result: (Boolean, Any?) -> Unit
    ) {
        if(ad != null) {
            // 이미 로드된 광고가 있으면 바로 리턴
            loadListener?.onAdLoadSuccess(this@FlutterDaroRewardAd, ad, null)
            result(true, null)
            return
        }
        if (loader == null) {
            setupLoader()
        }

        loader?.let {
            try {
                loadAdInternal(it, autoShow, listener, result)
            } catch (e: Exception) {
                result(false, mapOf(
                    "code" to 1011,
                    "message" to "Failed to load ad",
                    "details" to e.message,
                ))
                destroy()
            }
        } ?: result(false, mapOf(
            "code" to 1011,
            "message" to "Failed to load ad",
            "details" to "Failed to create loader",
        ))
    }

    // 공통 표시 로직
    fun showAd(
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        ad?.let { currentAd ->
            try {
                showAdInternal(currentAd, listener, result)
            } catch (e: Exception) {
                result(false, mapOf(
                    "code" to 1012,
                    "message" to "Failed to show ad",
                    "details" to e.message,
                ))
                destroy()
            }
        } ?: run {
            loadAd(autoShow = true, listener = listener, result = result)
        }
    }

    // 공통 해제 로직
    fun destroy() {
        ad?.let { destroyAd(it) }
        ad = null
        loader = null
        loadListener = null
    }
}

// 전면 광고 클래스
class FlutterDaroInterstitialAd(
    context: Context,
    adUnit: String,
    placement: String? = null,
    loadListener: FlutterDaroRewardAdLoadListener? = null
) : FlutterDaroRewardAd(context, adUnit, placement, loadListener) {

    override fun createLoader(): DaroInterstitialAdLoader {
        val unit = DaroInterstitialAdUnit(
            key = adUnit,
            placement = placement ?: ""
        )
        return DaroInterstitialAdLoader(
            context = context,
            adUnit = unit
        )
    }

    override fun loadAdInternal(
        loader: Any,
        autoShow: Boolean,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (loader as? DaroInterstitialAdLoader)?.setListener(object : DaroInterstitialAdLoaderListener {
            override fun onAdLoadSuccess(ad: DaroInterstitialAd, adInfo: DaroAdInfo) {
                this@FlutterDaroInterstitialAd.ad = ad
                loadListener?.onAdLoadSuccess(this@FlutterDaroInterstitialAd, ad, adInfo)
                
                if (autoShow) {
                    showAd(listener) { success, error ->
                        result(success, error)
                    }
                } else {
                    result(true, null)
                }
            }

            override fun onAdLoadFail(error: DaroAdLoadError) {
                loadListener?.onAdLoadFail(error)
                result(false, mapOf(
                    "code" to error.code,
                    "message" to error.message
                ))
            }
        })
        (loader as? DaroInterstitialAdLoader)?.load()
    }

    override fun showAdInternal(
        ad: Any,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (ad as? DaroInterstitialAd)?.let { interstitialAd ->
            interstitialAd.setListener(object : DaroInterstitialAdListener {
                override fun onAdImpression(adInfo: DaroAdInfo) {
                    listener?.onAdImpression(adInfo)
                }

                override fun onAdClicked(adInfo: DaroAdInfo) {
                    listener?.onAdClicked(adInfo)
                }

                override fun onShown(adInfo: DaroAdInfo) {
                    listener?.onShown(adInfo)
                }

                override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
                    listener?.onFailedToShow(adInfo, error)
                    destroy()
                    result(false, mapOf(
                        "code" to -1,//error.code,
                        "message" to error.message
                    ))
                }

                override fun onDismiss(adInfo: DaroAdInfo) {
                    listener?.onDismiss(adInfo)
                    destroy()
                    result(true, null)
                }
            })
            (context as? Activity)?.let { currentActivity ->
                interstitialAd.show(activity = currentActivity)
            } ?: {
                result(false, mapOf(
                    "code" to 1013,
                    "message" to "No activity available to show ad"
                ))
            }
        }
    }

    override fun destroyAd(ad: Any?) {
        (ad as? DaroInterstitialAd)?.destroy()
    }
}

// 리워드 비디오 광고 클래스
class FlutterDaroRewardedVideoAd(
    context: Context,
    adUnit: String,
    placement: String? = null,
    loadListener: FlutterDaroRewardAdLoadListener? = null
) : FlutterDaroRewardAd(context, adUnit, placement, loadListener) {

    override fun createLoader(): DaroRewardedAdLoader {
        val unit = DaroRewardedAdUnit(
            key = adUnit,
            placement = placement ?: ""
        )
        return DaroRewardedAdLoader(
            context = context,
            adUnit = unit
        )
    }

    override fun loadAdInternal(
        loader: Any,
        autoShow: Boolean,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (loader as? DaroRewardedAdLoader)?.setListener(object : DaroRewardedAdLoaderListener {
            override fun onAdLoadSuccess(ad: DaroRewardedAd, adInfo: DaroAdInfo) {
                this@FlutterDaroRewardedVideoAd.ad = ad
                loadListener?.onAdLoadSuccess(this@FlutterDaroRewardedVideoAd, ad, adInfo)
                
                if (autoShow) {
                    showAd(listener) { success, error ->
                        result(success, error)
                    }
                } else {
                    result(true, null)
                }
            }

            override fun onAdLoadFail(error: DaroAdLoadError) {
                loadListener?.onAdLoadFail(error)
                destroy()
                result(false, mapOf(
                    "code" to error.code,
                    "message" to error.message
                ))
            }
        })
        (loader as? DaroRewardedAdLoader)?.load()
    }

    override fun showAdInternal(
        ad: Any,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (ad as? DaroRewardedAd)?.let { rewardedAd ->
            var reward: DaroRewardedAd.DaroRewardedItem? = null
            rewardedAd.setListener(object : DaroRewardedAdListener {
                override fun onAdImpression(adInfo: DaroAdInfo) {
                    listener?.onAdImpression(adInfo)
                }

                override fun onAdClicked(adInfo: DaroAdInfo) {
                    listener?.onAdClicked(adInfo)
                }

                override fun onShown(adInfo: DaroAdInfo) {
                    listener?.onShown(adInfo)
                }

                override fun onEarnedReward(adInfo: DaroAdInfo, rewardItem: DaroRewardedAd.DaroRewardedItem) {
                    reward = rewardItem
                    listener?.onRewarded(adInfo, rewardItem)
                }

                override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
                    listener?.onFailedToShow(adInfo, error)
                    destroy()
                    result(false, mapOf(
                        "code" to -1,//error.code,
                        "message" to error.message
                    ))
                }

                override fun onDismiss(adInfo: DaroAdInfo) {
                    listener?.onDismiss(adInfo)
                    destroy()
                    result(reward != null, null)
                }
            })
            (context as? Activity)?.let { currentActivity ->
                rewardedAd.show(activity = currentActivity)
            } ?: {
                result(false, mapOf(
                    "code" to 1013,
                    "message" to "No activity available to show ad"
                ))
            }
        }
    }

    override fun destroyAd(ad: Any?) {
        (ad as? DaroRewardedAd)?.destroy()
    }
}

// 팝업 광고 클래스
class FlutterDaroPopupAd(
    context: Context,
    adUnit: String,
    placement: String? = null,
    loadListener: FlutterDaroRewardAdLoadListener? = null,
    private val options: Map<*, *>?,
) : FlutterDaroRewardAd(context, adUnit, placement, loadListener) {

    private fun createOptions(map: Map<*, *>?): DaroLightPopupAdOptions {
        if(map == null || map.isEmpty()) return DaroLightPopupAdOptions()
        fun valueToColorInt(value: Any?, defaultValue: String): Int {
            return when(value) {
                is Long -> {
                    val a = ((value shr 24) and 0xFF).toInt()
                    val r = ((value shr 16) and 0xFF).toInt()
                    val g = ((value shr 8) and 0xFF).toInt()
                    val b = (value and 0xFF).toInt()
                    Color.argb(a, r, g, b)
                }
                is Int -> {
                    val a = ((value shr 24) and 0xFF)
                    val r = ((value shr 16) and 0xFF)
                    val g = ((value shr 8) and 0xFF)
                    val b = (value and 0xFF)
                    Color.argb(a, r, g, b)
                }
                is String -> value.toColorInt()
                else -> defaultValue.toColorInt()
            }
        }
        return DaroLightPopupAdOptions(
            backgroundColor = valueToColorInt(map["backgroundColor"], "#B2121416"),
            containerColor = valueToColorInt(map["containerColor"], "#121416"),
            adMarkLabelTextColor = valueToColorInt(map["adMarkLabelTextColor"], "#F7FAFF"),
            adMarkLabelBackgroundColor = valueToColorInt(map["adMarkLabelBackgroundColor"], "#3E434F"),
            titleColor = valueToColorInt(map["titleColor"], "#F7FAFF"),
            bodyColor = valueToColorInt(map["bodyColor"], "#B6BECC"),
            ctaBackgroundColor = valueToColorInt(map["ctaBackgroundColor"], "#EB2640"),
            ctaTextColor= valueToColorInt(map["ctaTextColor"], "#FFFFFF"),
            closeButtonText = map["closeButtonText"] as? String ?: "Close", 
            closeButtonColor = valueToColorInt(map["closeButtonColor"], "#F7FAFF"),
        )
    }

    override fun createLoader(): DaroLightPopupAdLoader {
        val unit = DaroLightPopupAdUnit(
            key = adUnit,
            placement = placement ?: "",
            options = createOptions(options)
        )
        return DaroLightPopupAdLoader(
            context = context,
            adUnit = unit
        )
    }

    override fun loadAdInternal(
        loader: Any,
        autoShow: Boolean,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (loader as? DaroLightPopupAdLoader)?.setListener(object : DaroLightPopupAdLoaderListener {
            override fun onAdLoadSuccess(ad: DaroLightPopupAd, adInfo: DaroAdInfo) {
                this@FlutterDaroPopupAd.ad = ad
                loadListener?.onAdLoadSuccess(this@FlutterDaroPopupAd, ad, adInfo)
                
                if (autoShow) {
                    showAd(listener) { success, error ->
                        result(success, error)
                    }
                } else {
                    result(true, null)
                }
            }

            override fun onAdLoadFail(error: DaroAdLoadError) {
                loadListener?.onAdLoadFail(error)
                destroy()
                result(false, mapOf(
                    "code" to error.code,
                    "message" to error.message
                ))
            }
        })
        (loader as? DaroLightPopupAdLoader)?.load()
    }

    override fun showAdInternal(
        ad: Any,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (ad as? DaroLightPopupAd)?.let { popupAd ->
            popupAd.setListener(object : DaroLightPopupAdListener {
                override fun onAdImpression(adInfo: DaroAdInfo) {
                    listener?.onAdImpression(adInfo)
                }

                override fun onAdClicked(adInfo: DaroAdInfo) {
                    listener?.onAdClicked(adInfo)
                }

                override fun onShown(adInfo: DaroAdInfo) {
                    listener?.onShown(adInfo)
                }

                override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
                    listener?.onFailedToShow(adInfo, error)
                    destroy()
                    result(false, mapOf(
                        "code" to -1,//error.code,
                        "message" to error.message
                    ))
                }

                override fun onDismiss(adInfo: DaroAdInfo) {
                    listener?.onDismiss(adInfo)
                    destroy()
                    result(true, null)
                }
            })
            (context as? Activity)?.let { currentActivity ->
                popupAd.show(activity = currentActivity)
            } ?: {
                result(false, mapOf(
                    "code" to 1013,
                    "message" to "No activity available to show ad"
                ))
            }
        }
    }

    override fun destroyAd(ad: Any?) {
        (ad as? DaroLightPopupAd)?.destroy()
    }
}

// 앱 오프닝 광고 클래스
class FlutterDaroOpeningAd(
    context: Context,
    adUnit: String,
    placement: String? = null,
    loadListener: FlutterDaroRewardAdLoadListener? = null
) : FlutterDaroRewardAd(context, adUnit, placement, loadListener) {

    override fun createLoader(): DaroAppOpenAdLoader {
        val unit = DaroAppOpenAdUnit(
            key = adUnit,
            placement = placement ?: ""
        )
        return DaroAppOpenAdLoader(
            context = context,
            adUnit = unit
        )
    }

    override fun loadAdInternal(
        loader: Any,
        autoShow: Boolean,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (loader as? DaroAppOpenAdLoader)?.setListener(object : DaroAppOpenAdLoaderListener {
            override fun onAdLoadSuccess(ad: DaroAppOpenAd, adInfo: DaroAdInfo) {
                this@FlutterDaroOpeningAd.ad = ad
                loadListener?.onAdLoadSuccess(this@FlutterDaroOpeningAd, ad, adInfo)
                
                if (autoShow) {
                    showAd(listener) { success, error ->
                        result(success, error)
                    }
                } else {
                    result(true, null)
                }
            }

            override fun onAdLoadFail(error: DaroAdLoadError) {
                loadListener?.onAdLoadFail(error)
                destroy()
                result(false, mapOf(
                    "code" to error.code,
                    "message" to error.message
                ))
            }
        })
        (loader as? DaroAppOpenAdLoader)?.load()
    }

    override fun showAdInternal(
        ad: Any,
        listener: FlutterDaroRewardAdListener?,
        result: (Boolean, Any?) -> Unit
    ) {
        (ad as? DaroAppOpenAd)?.let { appOpenAd ->
            appOpenAd.setListener(object : DaroAppOpenAdListener {
                override fun onAdImpression(adInfo: DaroAdInfo) {
                    listener?.onAdImpression(adInfo)
                }

                override fun onAdClicked(adInfo: DaroAdInfo) {
                    listener?.onAdClicked(adInfo)
                }

                override fun onShown(adInfo: DaroAdInfo) {
                    listener?.onShown(adInfo)
                }

                override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
                    listener?.onFailedToShow(adInfo, error)
                    destroy()
                    result(false, mapOf(
                        "code" to -1,//error.code,
                        "message" to error.message
                    ))
                }

                override fun onDismiss(adInfo: DaroAdInfo) {
                    listener?.onDismiss(adInfo)
                    destroy()
                    result(true, null)
                }
            })

            (context as? Activity)?.let { currentActivity ->
                appOpenAd.show(activity = currentActivity)
            } ?: {
                result(false, mapOf(
                    "code" to 1013,
                    "message" to "No activity available to show ad"
                ))
            }
        }
    }

    override fun destroyAd(ad: Any?) {
        (ad as? DaroAppOpenAd)?.destroy()
    }
}